File: services/activity-service/app/schemas/activity.py
# activity.py

from pydantic import BaseModel, model_validator, field_validator
from datetime import datetime, timezone
from uuid import UUID
from typing import Literal, Optional


def _to_naive_utc(dt: datetime) -> datetime:
    if dt.tzinfo is not None:
        return dt.astimezone(timezone.utc).replace(tzinfo=None)
    return dt


# ── медиа ─────────────────────────────────────────────────────────────────────

class MediaItem(BaseModel):
    url: str
    type: Literal["photo", "video"]


class MediaItemRead(BaseModel):
    id: UUID
    url: str
    type: Literal["photo", "video"]

    class Config:
        from_attributes = True


# ── активность ────────────────────────────────────────────────────────────────

class ActivityCreate(BaseModel):
    type: Literal["event", "meeting"]
    title: str
    description: Optional[str] = None
    latitude: float
    longitude: float
    starts_at: datetime
    expires_at: datetime
    max_participants: Optional[int] = None
    media: list[MediaItem] = []  # список {url, type} которые вернул media-service

    @field_validator("starts_at", "expires_at", mode="after")
    @classmethod
    def normalize_dt(cls, v: datetime) -> datetime:
        return _to_naive_utc(v)


class ActivityUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    starts_at: Optional[datetime] = None
    expires_at: Optional[datetime] = None
    max_participants: Optional[int] = None

    @field_validator("starts_at", "expires_at", mode="after")
    @classmethod
    def normalize_dt(cls, v: Optional[datetime]) -> Optional[datetime]:
        if v is None:
            return v
        return _to_naive_utc(v)

    @model_validator(mode='after')
    def coords_both_or_neither(self):
        has_lat = self.latitude is not None
        has_lng = self.longitude is not None
        if has_lat != has_lng:
            raise ValueError('latitude и longitude должны передаваться вместе')
        return self


class ActivityRead(BaseModel):
    id: UUID
    type: Literal["event", "meeting"]
    creator_id: UUID
    title: str
    description: Optional[str] = None
    latitude: float
    longitude: float
    starts_at: datetime
    expires_at: datetime
    max_participants: Optional[int] = None
    media: list[MediaItemRead] = []

    class Config:
        from_attributes = True

File: mobile/lib/map/activity.dart
// lib/map/activity.dart

class MediaItem {
  final String url;
  final String type; // "photo" | "video"

  const MediaItem({required this.url, required this.type});

  factory MediaItem.fromJson(Map<String, dynamic> json) =>
      MediaItem(url: json['url'] as String, type: json['type'] as String);

  Map<String, dynamic> toJson() => {'url': url, 'type': type};
}

class Activity {
  final String id;
  final String type;
  final String creatorId; // новое
  final String authorName; // новое
  final String title;
  final String? description;
  final double latitude;
  final double longitude;
  final DateTime startsAt;
  final DateTime expiresAt;
  final int? maxParticipants;
  final List<MediaItem> media;

  const Activity({
    required this.id,
    required this.type,
    required this.creatorId, // новое
    this.authorName = '', // новое
    required this.title,
    this.description,
    this.maxParticipants,
    required this.latitude,
    required this.longitude,
    required this.startsAt,
    required this.expiresAt,
    this.media = const [],
  });

  Activity copyWith({String? authorName}) => Activity(
    id: id,
    type: type,
    creatorId: creatorId,
    authorName: authorName ?? this.authorName,
    title: title,
    description: description,
    latitude: latitude,
    longitude: longitude,
    startsAt: startsAt,
    expiresAt: expiresAt,
    maxParticipants: maxParticipants,
    media: media,
  );

  factory Activity.fromJson(Map<String, dynamic> json) => Activity(
    id: json['id'] as String,
    type: json['type'] as String,
    creatorId: json['creator_id'] as String, // новое
    title: json['title'] as String,
    description: json['description'] as String?,
    latitude: (json['latitude'] as num).toDouble(),
    longitude: (json['longitude'] as num).toDouble(),
    startsAt: DateTime.parse(json['starts_at'] as String),
    expiresAt: DateTime.parse(json['expires_at'] as String),
    maxParticipants: json['max_participants'] as int?,
    media: (json['media'] as List<dynamic>? ?? [])
        .map((e) => MediaItem.fromJson(e as Map<String, dynamic>))
        .toList(),
  );
}
